runtime/vm/JFRWriter.hpp

```diff
@@ -0,0 +1,167 @@
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+#if !defined(JFRWRITER_HPP_)
+#define JFRWRITER_HPP_
+
+#include <stdlib.h>
+
+#include "j9cfg.h"
+#include "j9.h"
+#include "vm_api.h"
+
+extern "C" {
+class VM_JFRWriter {
+	/*
+	 * Data members
+	 */
+private:
+
+protected:
+
+public:
+	static constexpr const char* METADATA_BLOB_FILE_EMVVAR = "OPENJ9_METADATA_BLOB_FILE_PATH";
+
+	static constexpr const char* DEFAULT_JFR_FILE_NAME = "defaultJ9recording.jfr";
+
+
+	/*
+	 * Function members
+	 */
+private:
+	static void
+	closeJFRFile(J9JavaVM *vm)
+	{
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		if (-1 != vm->jfrState.blobFileDescriptor) {
+			j9file_close(vm->jfrState.blobFileDescriptor);
+		}
+	}
+
+	static bool
+	loadJFRMetadataBlob(J9JavaVM *vm)
+	{
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		I_64 fileLength = 0;
+		IDATA fileDescriptor = 0;
+		bool result = false;
+
+		//TODO need a better way of doing this
+		char *blobDir = getenv(METADATA_BLOB_FILE_EMVVAR);
+		if (NULL == blobDir) {
+			goto done;
+		}
+
+		fileLength = j9file_length(blobDir);
+		/* Restrict file size to < 2G */
+		if (fileLength > J9CONST64(0x7FFFFFFF)) {
+			goto done;
+		}
+
+		if ((fileDescriptor = j9file_open(blobDir, EsOpenRead, 0)) == -1) {
+			goto done;
+		}
+
+		vm->jfrState.metaDataBlobFileSize = fileLength;
+
+		vm->jfrState.metaDataBlobFile = (U_8*) j9mem_allocate_memory(vm->jfrState.metaDataBlobFileSize, OMRMEM_CATEGORY_VM);
+		if (NULL == vm->jfrState.metaDataBlobFile) {
+			goto done;
+		}
+
+		if (vm->jfrState.metaDataBlobFileSize != (U_64)(I_64)j9file_read(fileDescriptor, vm->jfrState.metaDataBlobFile, vm->jfrState.metaDataBlobFileSize)) {
+			goto done;
+		}
+
+		j9file_close(fileDescriptor);
+		result = true;
+
+done:
+		return result;
+	}
+
+	static bool
+	openJFRFile(J9JavaVM *vm)
+	{
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		bool result = true;
+		I_64 fileDescriptor = j9file_open(vm->jfrState.jfrFileName, EsOpenWrite | EsOpenCreate | EsOpenTruncate , 0666);
+
+		if (-1 == fileDescriptor) {
+			result = false;
+			goto done;
```

Comment:
This means the -1 will not be written to the descriptor field, which will makes the close code above will fail the check.